Added Kotlin support for Bukkit, Spigot and PaperMC projects.

Lasslos opened the pull request:
I added Kotlin support for Bukkit, Spigot and PaperMC projects. 

Plus, Kotlin support is easily extendable on any other platform. Just declare in your module that you support Kotlin `override val supportedLanguages: List<LanguageType> = listOf(LanguageType.JAVA, LanguageType.KOTLIN)
` and your ready to go. In your builder, you will be provided a LanguageType instance that can be used to acctually support multiple languages.

RedNesto commented on src/main/kotlin/language/Language.kt:
```diff
@@ -0,0 +1,9 @@
+
+import javax.swing.Icon
+
+abstract class Language {
```
I do not like having both `Language` and `LanguageType`, especially because `Language` already exists in the platform, which is somewhat confusing.

If these classes are only meant to be used in the project creator they should be located in `com.demonwav.mcdev.creator`. The whole thing could be simplified to be a `com.demonwav.mcdev.creator.CreatorLanguage` enum like
```kotlin
enum class CreatorLanguage(val displayName: String, val icon: Icon, val fileEnding: String) {
    JAVA("Java", LanguageAssets.JAVA_ICON, ".java"),
    KOTLIN("Kotlin", LanguageAssets.KOTLIN_ICON, ".kt");
} 
```

RedNesto commented on src/main/kotlin/creator/LanguageChooserWizardStep.kt:
```diff
@@ -0,0 +1,50 @@
+import com.intellij.ide.util.projectWizard.ModuleWizardStep
+import javax.swing.*
+
+class LanguageChooserWizardStep(private val creator: MinecraftProjectCreator) : ModuleWizardStep() {
```
Having a wizard step whose single purpose is to select a language is really overkill.
I would add a simple ComboBox to `BuildSystemWizardStep` instead.

RedNesto commented on src/main/kotlin/platform/bukkit/creator/BukkitTemplate.kt:
```diff
@@ -34,29 +36,42 @@ object BukkitTemplate : BaseTemplate() {
 
-        return project.applyTemplate(BUKKIT_MAIN_CLASS_TEMPLATE, props)
+        return project.applyTemplate(
+            when (languageType) {
+                LanguageType.JAVA -> BUKKIT_JAVA_MAIN_CLASS_TEMPLATE
+                LanguageType.KOTLIN -> BUKKIT_KOTLIN_MAIN_CLASS_TEMPLATE
```
Wouldn't it be simpler to have a method `MinecraftTemplates.getVariant(BUKKIT_MAIN_CLASS_TEMPLATE, language = languageType)` where `BUKKIT_MAIN_CLASS_TEMPLATE = "Bukkit Main Class"` and it would append `language.fileEnding` ?

RedNesto commented on src/main/resources/fileTemplates/j2ee/bukkit/Bukkit pom.xml.ft:
```diff
@@ -43,6 +48,7 @@
           </execution>
         </executions>
       </plugin>
+
```
Please remove this change.

RedNesto commented on src/main/kotlin/platform/AbstractModuleType.kt:
```diff
@@ -36,6 +37,8 @@ abstract class AbstractModuleType<out T : AbstractModule>(val groupId: String, v
 
     abstract val id: String
 
+    abstract val supportedLanguages: List<LanguageType>
```
I don't think this information should be in ModuleTypes, but rather in ProjectConfig. I don't see  any other use for this information other than for the project creator.

RedNesto commented on src/main/kotlin/creator/buildsystem/DirectorySet.kt:
```diff
@@ -20,10 +21,11 @@ data class DirectorySet(
-        fun create(dir: Path): DirectorySet {
-            val sourceDirectory = dir.resolve("src/main/java")
+        fun create(dir: Path, languageType: LanguageType = LanguageType.JAVA): DirectorySet {
+            val sourceDirectory = dir.resolve("src/main/" + if (languageType == LanguageType.JAVA) "java" else "kotlin")
             val resourceDirectory = dir.resolve("src/main/resources")
-            val testSourceDirectory = dir.resolve("src/test/java")
+            val testSourceDirectory =
+                dir.resolve("src/test/" + if (languageType == LanguageType.JAVA) "java" else "kotlin")
```
You should introduce a new variable instead of having using the same if expression twice.

RedNesto commented on src/main/kotlin/creator/buildsystem/gradle/GradleBuildSystem.kt:
```diff
@@ -20,7 +20,7 @@ import com.demonwav.mcdev.platform.PlatformType
 import com.intellij.openapi.module.Module
 import java.nio.file.Path
-import java.util.Locale
+import java.util.*
```
Please revert this change.

RedNesto commented on src/main/kotlin/platform/sponge/creator/SpongeProjectCreator.kt:
```diff
@@ -41,7 +32,7 @@ import com.intellij.psi.codeStyle.CodeStyleManager
 import java.nio.file.Path
 import java.nio.file.Paths
-import java.util.EnumSet
+import java.util.*
```
These imports changes are not necessary.

RedNesto commented on src/main/kotlin/platform/velocity/creator/VelocityProjectCreator.kt:
```diff
@@ -17,19 +17,8 @@ import com.demonwav.mcdev.creator.CreatorStep
-import com.demonwav.mcdev.creator.buildsystem.maven.MavenBuildSystem
-import com.demonwav.mcdev.creator.buildsystem.maven.MavenGitignoreStep
+import com.demonwav.mcdev.creator.buildsystem.gradle.*
+import com.demonwav.mcdev.creator.buildsystem.maven.*
```
Unnecessary changes.

RedNesto commented on src/main/kotlin/util/MinecraftTemplates.kt:
```diff
@@ -23,7 +23,7 @@ class MinecraftTemplates : FileTemplateGroupDescriptorFactory {
         FileTemplateGroupDescriptor("Bukkit", PlatformAssets.BUKKIT_ICON).let { bukkitGroup ->
             group.addTemplate(bukkitGroup)
-            bukkitGroup.addTemplate(FileTemplateDescriptor(BUKKIT_MAIN_CLASS_TEMPLATE))
+            bukkitGroup.addTemplate(FileTemplateDescriptor(BUKKIT_JAVA_MAIN_CLASS_TEMPLATE))
```
I don't see the Kotlin variant added anywhere.

As I mentioned in another comment it might be better to have a single base template name and register templates with all tge language variants supported.
We could have a `FileTemplateGroupDescriptor.addTemplate(name: String, vararg languages: LanguageType)` extension method simply appending the base `name` with the file languages extension. Or a `template` overload with the same arguments and logic (see the bottom of this file)

RedNesto reviewed (changes requested):
Overall this is a good start. Kind of an early review given I'd like to do a couple of changes to the wizard steps and ProjectConfigs along with some fixes, so expect some more conflicts.

Please rebase against `dev` before doing any more work on this PR because the base IDEA version was bumped to 2020.3.

I think you forgot to run the formatter and format checks (`./gradlew format`, `./gradlew check`, and see https://github.com/minecraft-dev/MinecraftDev#ide-setup)

Also please avoid committing unneeded changes like blank lines and star imports (this might be because you did not use the Ktlint formatter configuration.)

Lasslos commented:
Thanks for your detailed feedback, I'll soon edit my fork. Sorry that I changed so many unnecessary things like imports etc, that's because I've chosen optimise imports when I committed in Intellij.

RedNesto commented:
Your last commit doesn't look right. You should drop it and do a rebase against `dev` instead of a merge.

Lasslos commented:
I hope this is better now. This is the first time me contributing to an open source project so I'm not very experienced... Sorry about that.

RedNesto commented:
Ah not quite there, but don't worry I messed up many of my PRs with that ^^

I can probably do it for you when I get enough time.

Lasslos commented:
That would be great.

RedNesto commented:
Done, I also fixed the various formatting problems.

The simplest way for you to get the new commits is to pull your origin/dev branch, or reset against it after a fetch.

Lasslos commented:
Thank you, I will focus on changing the actual code. And btw, I am using GitHub Desktop, so pulling and pushing is no problem at all.

RedNesto commented:
You broke it again...

Lasslos commented:
I changed a lot of things now. But, there are still a lot of files changed that I didn't want to change. Could it be that something went wrong when rebasing the project?

RedNesto commented:
The reason it got bad again is because you did not reset/recreate the branch like I told you, so you pulled my changes on top of the previous broken commits. I'll fix it once more and this time you delete you local branch and pull the remote fresh.

Lasslos commented:
Oh, now I understand. Sorry about that. When I read your message "the best way to get the new commits" I thought oh nice at least I know how to do that :/. Apparently I didn't. But thank you that you fix it for me!

RedNesto commented:
Here it is, with little improvements to fix one of my comment, and to fix the default language checkbox selection.

Also remember to properly format the files before committing, see https://github.com/minecraft-dev/MinecraftDev#ide-setup

Do you plan to add templates for other platforms and/or Gradle's Kotlin DSL ?

Lasslos commented:
I was planning to do this, but I'm not an expert in these platforms, so I have to see which Buildsscripts are easy to change. Is every Platform JVM-Only?

RedNesto commented:
Yes all supported platforms run on the JVM. If you don't feel like doing it don't worry, I can handle most of the other platforms without much trouble (except fabric)

As it is right now this PR could be merged soon-ish (but it will sit in nightly for what I expect to be a long time) so we have a lot of time to smooth things out if needed.

Earthcomputer commented on src/main/kotlin/platform/bukkit/creator/BukkitTemplate.kt:
```diff
@@ -33,29 +35,43 @@ object BukkitTemplate : BaseTemplate() {
+        val properties = BasicMavenStep.pluginVersions.toMutableMap()
+        if (language == CreatorLanguage.KOTLIN) {
+            properties["USE_KOTLIN"] = "true"
+        }
```
Couldn't this be something more like `properties["USE_${language.name}"] = "true"`?

Earthcomputer commented on src/main/kotlin/platform/bukkit/creator/BukkitTemplate.kt:
```diff
@@ -33,29 +35,43 @@ object BukkitTemplate : BaseTemplate() {
+                CreatorLanguage.JAVA -> "USE_JAVA"
+                CreatorLanguage.KOTLIN -> "USE_KOTLIN"
+            }
+        ] = "true"
```
same here

Lasslos replied:
Yes. Can you make a PR? Otherwise I'll change it lateron

Earthcomputer replied:
I'd rather leave it to you xD

RedGrapefruit09 commented:
What's the status on this PR?

Lasslos commented:
Idk, I fixed everything requested. I guess everyone (including me) kinnda forgot about it ¯\_(ツ)_/¯

Lasslos commented:
Hi, its 2022 now and I just wanted to ask how this PR is doing. I will (hopefully) have a lot of time to work on my projects in a couple weeks. Is there anything that **needs to be done?**

RedNesto commented:
What is needed is kotlin templates for all the other platforms, and maybe figure out what's the best way to have kotlin in the runtime (plugin/mod dependency probably, unsure yet but this can wait)

I wanted to do the other templates in september but I got unexpectedly busy. Will try to get some time to at least add the Sponge and Forge templates.

And please stop making useless merge commits, they're unnecessary noise ^^

Lasslos commented:
Hi, I tried to work on the templates+integration today and realized, that somehow, my fork is broken. 

I cant get anything to work, everything says unresolved reference. 
![Screenshot from 2022-03-09 19-11-33](https://user-images.githubusercontent.com/81803114/157574665-cd942169-27a0-4109-b3e3-847528d72e62.png)

I double checked that everything is working on this version. I don't wanna bother you with this issue and will hopefully find a solution but if anyone had a similar problem and knows something about that any help would be appreciated.

Lasslos commented:
For some reason it is fixed now, i dont really understand Gradle so cant tell you why

DenWav commented on src/main/resources/fileTemplates/j2ee/bukkit/Bukkit Main Class.kt.ft:
```diff
@@ -0,0 +1,14 @@
+
+import org.bukkit.plugin.java.JavaPlugin
+
+class ${CLASS_NAME}: JavaPlugin() {
```
Should be a space before the `:`.

DenWav commented on src/main/resources/fileTemplates/j2ee/bungeecord/BungeeCord Main Class.kt.ft:
```diff
@@ -0,0 +1,14 @@
+
+import net.md_5.bungee.api.plugin.Plugin
+
+class ${CLASS_NAME}: Plugin() {
```
Should be a space before the `:`.

sttephen commented:
> For some reason it is fixed now, i dont really understand Gradle so cant tell you why

Any update on this mate? You would be doing the world a huge favour 🙏

Lasslos commented:
Hi there, sorry for the delay. 
#### State/History
I changed some code that added Kotlin support for Bukkit, Spigot and PaperMC two years ago. Admittedly, it wasn't good code by any means (I was quite unexperienced and still am, but I think I've learned :smile: ), but it works. Now after some changes, we have reached an acceptable state for Bukkit, Spigot and Paper (as far as I can remember, however I didn't look at the code again as I am writing this)

#### Open things

- Review Old Code - I (or anyone) should definitely review all the code that's been added
- Check if the way Kotlin is implemented makes sense -  I just remember not really knowing what I am doing, so maybe we should review if what I have added back then makes sense where it is, or if it should follow a different design pattern
- Test - The code needs to be tested extensively, especially with newer versions of Minecraft as I don't know if everything still works the same
- (most important) Add Kotlin templates for all the other Project Types

To generate these Templates or to get an idea of how they should look like, you can create a Java Project and then use IntelliJs "Convert to Kotlin Project" function, as that usually outputs a working gradle build script. This is also the workaround that can be used to code plugins in Kotlin, doesnt work great tho.

I have not had huge interest in this project recently, but I am happy to pick it up again with a few contributes, which would especially help with adding Kotlin for the other project types.

Is there any interest in this being actively worked on again? I'd appreciate your feedback as well as your points that have to be done in your opinion.

Lasslos commented:
I just looked into the code and saw that there were massive improvements all over the plugin, making it less work to redo the work I did earlier. I've created the branch [kotlin](https://github.com/Lasslos/MinecraftDev/tree/feature/kotlin), which is up to date with this project's dev branch, and moved the old code to [olddev](https://github.com/Lasslos/MinecraftDev/tree/olddev).

Because there are no more commits in dev, this pull request automatically closed, but we can continue any discussion here. I would suggest creating a new PR and merging it immediately as soon as we are all convinced the feature is ready, but I am open to any other suggestions as well doing some git magic to reopen this PR and renaming some branches, but I don't think that's worth the trouble.
